[WIP][JIT] __str__ support for c++ bound classes

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32456 [WIP][JIT] __str__ support for c++ bound classes**

Differential Revision: [D19504482](https://our.internmc.facebook.com/intern/diff/D19504482)

Commit:
[JIT] __str__ support for c++ bound classes

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 3f4262c6:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=3f4262c6937846b050fbd19985c19d031e51a702).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4308568) pytorch_linux_xenial_py2_7_9_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89059758) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89059758))
<details>
<summary>
<code>Jan 21 21:52:02 RuntimeError: test_quantization failed!</code>
</summary>

```
Jan 21 21:52:02 Ran 36 tests in 52.514s 
Jan 21 21:52:02  
Jan 21 21:52:02 FAILED (errors=2, skipped=2) 
Jan 21 21:52:02  
Jan 21 21:52:02 Generating XML reports... 
Jan 21 21:52:02 Traceback (most recent call last): 
Jan 21 21:52:02   File "test/run_test.py", line 456, in <module> 
Jan 21 21:52:02     main() 
Jan 21 21:52:02   File "test/run_test.py", line 449, in main 
Jan 21 21:52:02     raise RuntimeError(message) 
Jan 21 21:52:02 RuntimeError: test_quantization failed! 
Jan 21 21:52:02 + cleanup 
Jan 21 21:52:02 + retcode=1 
Jan 21 21:52:02 + set +x 
Jan 21 21:52:02 =================== sccache compilation log =================== 
Jan 21 21:52:02 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 21 21:52:02 Compile requests                 10 
Jan 21 21:52:02 Compile requests executed         8 
Jan 21 21:52:02 Cache hits                        2 
Jan 21 21:52:02 Cache misses                      6 
Jan 21 21:52:02 Cache timeouts                    0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

eellison commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -748,11 +748,20 @@ void initJitScriptBindings(PyObject* module) {
+          method->run(s);
+          return s.back().toString()->string();
+        }
+        throw std::runtime_error("__str__ is not defined!");
```
It's somewhat niche, but in theory you should be raising `NotImplemented` error here. For example, in `a + b` if `__add__` isn't defined on `a` `__radd__` is called on b. Throwing something other than `NotImplemented`  in `__add__` will prevent this behavior. I don't if there are any other places where there's a similar hierarchy with `__str__`.

Commit:
Update on "[JIT] __str__ support for c++ bound classes"


Differential Revision: [D19504482](https://our.internmc.facebook.com/intern/diff/D19504482)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] __str__ support for c++ bound classes"


Differential Revision: [D19504482](https://our.internmc.facebook.com/intern/diff/D19504482)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] __str__ support for c++ bound classes"


Differential Revision: [D19504482](https://our.internmc.facebook.com/intern/diff/D19504482)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] __str__ support for c++ bound classes"


Differential Revision: [D19504482](https://our.internmc.facebook.com/intern/diff/D19504482)

[ghstack-poisoned]
